### PR TITLE
Close search bar in Connect on blur

### PR DESCRIPTION
**Pull request by ravicious**
Adding `onBlur` to the search input fixes two scenarios:

* Pressing the shortcut to open a new terminal tab while the search bar is open now automatically closes the search bar.
* It is now possible to tab out of the search bar.
   * I don't 100% understand what the problem was before. Matter of fact is that pressing Tab while the input was focused did result on the input receiving a blur event, but the input was receiving focus back again. Since we had `onFocus` on the overall search bar container and not the input itself, I imagine it might have went like this:
      * You click on the search bar.
      * `onFocus` on the search bar container gets called.
      * `open` gets called, rendering a new input with `autoFocus`.
      * The new input gets focused.
      * Pressing Tab at this point blurs the input and shifts focus to another element within the search bar.
      * `onFocus` of the search bar container gets triggered again.
   * I haven't verified this though.

### CODE REVIEW

File: web/packages/teleterm/src/ui/Search/SearchBar.test.tsx
Comment (ravicious) [inline web/packages/teleterm/src/ui/Search/SearchBar.test.tsx:326]: In the upcoming commits I'm going to add `closeWithoutRestoringFocus`. I didn't want to have three different `close` variants and I didn't want to have `close({restoreFocus: boolean, resetInput: boolean})` either. So I decided to get rid of `closeAndResetInput` as it was used only in a single specific situation which already also called `resetInput` anyway.

File: web/packages/teleterm/src/ui/Search/SearchBar.tsx
Comment (ravicious) [inline web/packages/teleterm/src/ui/Search/SearchBar.tsx:100]: This is an unorthodox name, but I don't like listeners being called like `onBlur` and `handleBlur` if they do something more complex because then you have no idea what's happening in them.

That's the same reason why I renamed the prop in the following piece of code from `onShowDetails` to `showErrorsInModal`:

https://github.com/gravitational/teleport/blob/5bf72af28a5f3fdacd09717a6b5bad0a08b2eb60/web/packages/teleterm/src/ui/Search/pickers/ActionPicker.tsx#L217-L220

File: web/packages/teleterm/src/ui/Search/SearchContext.tsx
Comment (ravicious) [inline web/packages/teleterm/src/ui/Search/SearchContext.tsx]: I wish I didn't have to implement it like this, I'd rather have `close` which doesn't restore focus and then another version of `close` which does. This way we wouldn't have to depend on some state interactions to achieve the effect we want.

But given that in all but one case we want to restore focus, I just couldn't find a way to do this without making the code unnecessarily complicated.

File: web/packages/teleterm/src/ui/Search/SearchContext.tsx
Comment (ravicious) [inline web/packages/teleterm/src/ui/Search/SearchContext.tsx:176]: Technically the function has nothing to do with event listeners, it just simply does not return the passed value if `isUserInteractionPaused` is true.

So we could have a more generic name but since this function would be used only for event listeners, I feel like `makeEventListener` is a better name than a generic one.

I tried coming up with something like `makePausableEventListener` but I didn't think of anything that I'd like.

**Review by avatus — APPROVED** [review]

File: web/packages/teleterm/src/ui/Search/SearchBar.tsx
Comment (ravicious) [reply web/packages/teleterm/src/ui/Search/SearchBar.tsx:100]: Hmm maybe the complexity of the handler is the wrong way to look at it. What I should've said is that if I'm writing a generic component that's going to be used in many different contexts, say a `<Button>` or `<Input>`, then `onClick`, `onBlur` accepted through props all make sense because `<Button>` doesn't care what the handler does.

But if I'm working on a specific component in a specific context, then IMHO giving specific names to handlers makes it easier to understand what's going on, as you described. Otherwise all I see is just `onBlur` or `handleBlur` and if I want to know what actually happens on blur, I have to go to the function _and_ I have to read through it because there's no variable name which would summarize what the function does.

Matter of fact, the `onInputValueChange` function exported from `SearchContext` has this problem too, I'm going to rename it in the next commit.

File: web/packages/teleterm/src/ui/Search/SearchContext.tsx
Comment (gzdunek) [reply web/packages/teleterm/src/ui/Search/SearchContext.tsx:80]: AFAIK `previouslyActive` can be for example an `SVGElement` that doesn't have focus method, so we want to run it only for `HTMLElement`. That's why I checked an instance of the element.  
https://stackoverflow.com/a/47198823

I think you can still have that simple mock in a test, but only revert the `instanceof` check?

File: web/packages/teleterm/src/ui/Search/SearchBar.tsx
Comment (gzdunek) [reply web/packages/teleterm/src/ui/Search/SearchBar.tsx:100]: I think it makes sense to use more specific names when possible. 
If I needed to run a few unrelated actions in a handler, then probably I would just call it `handleBlur`.

**Review by gzdunek — APPROVED** [review]
Looks good, but I think we should revert checking the instance of `previouslyActive`, so the app won't crash in the runtime.

File: web/packages/teleterm/src/ui/Search/SearchContext.tsx
Comment (ravicious) [reply web/packages/teleterm/src/ui/Search/SearchContext.tsx:80]: `SVGElement` extends `HTMLOrSVGElement` which has `focus`, though I've seen people mention what you said (https://github.com/Microsoft/TypeScript/issues/5901#issuecomment-161716750), but I suppose it has changed since 2015.

I cannot use the `instanceof` check together with a simple mock because the `instanceof` check is done in runtime and the object is not an instance of `HTMLElement`.

[MDN doesn't explain](https://developer.mozilla.org/en-US/docs/Web/API/Document/activeElement) how it's possible for `document.activeElement` to be an `Element` that doesn't have `focus` function. I assume this is just due to the internal browser engine hierarchy of DOM objects.

For now, I'll revert `previouslyActive` back to just `Element` to adhere to the browser API. Before calling `focus` I'll check if `previouslyActive` has the `focus` property which is a function. I think this is better than an `instanceof` check because after all we're not interested in the element being an instance of a specific class, we just want to call `focus` on it.

File: web/packages/teleterm/src/ui/Search/SearchBar.tsx
Comment (ravicious) [reply web/packages/teleterm/src/ui/Search/SearchBar.tsx:117]: Ah, good catch, I forgot to remove it when I was still prototyping since I originally called just `close` there. ;)

File: web/packages/teleterm/src/ui/Search/SearchContext.tsx
Comment (ravicious) [reply web/packages/teleterm/src/ui/Search/SearchContext.tsx:80]: I have to use a less type-safe check anyway because the current code won't pass CI until #25683 gets merged and I don't want to be blocked by it. I'll add a TODO comment to address it later.

**Review by ryanclark — APPROVED** [review]

**Pull request merged**
